### PR TITLE
Adding prostests analyser

### PR DESCRIPTION
**Pull request by Smoltbob**
I added a folder and tried to replicate the structure of KerasPretrained.
There is also a temporary script `test.py` that allows to test the model outside of mtriage.

### CODE REVIEW

**Comment by Smoltbob** [comment]
This is built using the trained model from the authors of the following repository. https://github.com/wondonghyeon/protest-detection-violence-estimation

**Comment by breezykermo** [comment]
Hi @Smoltbob, sorry for the delay. One basic update that is needed here to get the build passing is to lint the code using [black](https://github.com/psf/black). If you've pip installed the outer requirements.txt, there's a script to easily do this for all relevant files:
```bash
sh scripts/lint.sh
```

File: src/lib/analysers/ProtestsPretrained/requirements.txt
Comment (breezykermo) [inline src/lib/analysers/ProtestsPretrained/requirements.txt]: This should be 'torch'. When building a Docker image, mtriage collects all the requirements from components' requirements.txt files and installs them- and the correct way to install pytorch through pip is `pip install torch`.

Can you give a version as well? i.e. `torch==1.4`.

Is `torchvision` also required? If so it needs to also be listed here.

File: src/lib/analysers/ProtestsPretrained/requirements.txt
Comment (breezykermo) [inline src/lib/analysers/ProtestsPretrained/requirements.txt]: An explicit version for pillow would also be better. 

Mtriage doesn't deal _that_ well with incompatible versions between analysers yet... but it's in the works, and so explicit labelling in component requirements is ideal.

File: src/lib/analysers/ProtestsPretrained/core.py
Comment (breezykermo) [inline src/lib/analysers/ProtestsPretrained/core.py]: This setup looks really good in principle!

File: src/lib/analysers/ProtestsPretrained/core.py
Comment (breezykermo) [inline src/lib/analysers/ProtestsPretrained/core.py]: Can delete these two lines, as they're not being used anywhere.

**Review by breezykermo — CHANGES_REQUESTED** [review]
Thanks so much for this, looking really good! I've made a couple of comments inline, and a little later today I'll comment a longer writeup explaining how I'm going about creating a custom build of mtriage to test this analyser on real data. None of this testing stuff is documented yet, sorry, but hopefully you can then apply it in the YoloV3 PR you've made as well, and I can take it and turn it into more general documentation for other contributors. :)

**Comment by Smoltbob** [comment]
Thank you,
- I wrote this prototype requirement.txt and forgot to make it proper, sorry!
- I'll re-lint the script, I had done it at some point but I must not have redone it before pushing

File: src/lib/analysers/ProtestsPretrained/requirements.txt
Comment (Smoltbob) [reply src/lib/analysers/ProtestsPretrained/requirements.txt]: I added the main requirements, indeed including torchvision. I don't think the newest versions of torch & co are needed, I could test with other versions if needed.

File: src/lib/analysers/ProtestsPretrained/core.py
Comment (Smoltbob) [reply src/lib/analysers/ProtestsPretrained/core.py]: Removed!

**Comment by breezykermo** [comment]
Fantastic, thanks. I think this build may fail regardless due to something I need to fix in the core mtriage runtime. I'm going to post a long explanation here now about how you can test `ProtestsPretrained` on real data with a custom build.

**Comment by breezykermo** [comment]
## Downloading some example data

I'm going to create an mtriage config that selects some video frames from Youtube, so that we can run the ProtestsPretrained analyser on them and simulate a real-world mtriage run. Running a generic search for "tear gas + mexico" in a span of 15 days gives me ~460 results, which running on my 4 CPU computer takes about 15 minutes to download using mtriage, and amounts to ~1.5gb of videos (mtriage downloads them all very low res by default). 

If you don't have enough space or want to try on a smaller sample, just tweak the `uploaded_after` parameter to narrow down the range of dates.

I'm going to split my config into two parts, so that I don't have to redownload videos on every run while I'm debugging the ProtestsPretrained analyser.

Let's put this is 'data/download_vids.yml':

```yaml
folder: media/test_protests
select:
  name: Youtube
  config:
    search_term: tear gas + mexico
    uploaded_before: "2018-11-30T00:00:00Z"
    uploaded_after: "2018-11-15T00:00:00Z"
analyse:
  - name: Frames
```

Make sure you've followed the [youtube selector setup](https://github.com/forensic-architecture/mtriage/blob/master/docs/components/youtube.md) properly as well.

Now I can run this config using the regular build of mtriage (to do Youtube selecting and Frames) with the following command.

```bash
./mtriage run data/download_vids.yml --dev
```

This will download all the videos, and split them up into frames. 

## Testing ProtestsPretrained with a custom build

Create a 'whitelist.txt' file with the following lines:

```
ProtestsPretrained
```

This will allow us to build a stripped down version of mtriage with only the core library, and the dependencies for this component in development.

We can now build that with this command:

```bash
./mtriage dev build --tag protests --whitelist whitelist.txt
```

Now I have a build of mtriage with the ProtestsPretrained dependencies that I can experimentally run and debug.

When we run mtriage with this analyser, we want to use our custom build, to ensure that all of the dependencies specified in the ProtestsPretrained analyser are sufficient and correct:

Let's create a yml that operates on the data we downloaded in the previous step in 'data/test_protests.yml':

```yaml
folder: media/test_protests
elements_in:
  - Youtube/Frames
analyse:
  - name: ProtestsPretrained
    config:
      labels:
        - protest
```

Now we can test a build of mtriage that only has the ProtestsPretrained requirements installed with the following command:

```bash
./mtriage run data/test_protests.yml --tag protests --dev
```

**Comment by breezykermo** [comment]
I've walked through this setup, and I think at least more dependencies- `pandas` and `numpy`- are required.

**Comment by Smoltbob** [comment]
Thank you ! Let's test all this then ! Yes there is a numpy dependency, I hadn't included it thinking it was standard. I'll track the pandas, I think I can remove this dependency.

**Comment by Smoltbob** [comment]
I've tested on a full run with no issues!

**Comment by breezykermo** [comment]
Awesome thanks. Sorry for the slow pickup here. It's in the works

**Review by breezykermo — APPROVED** [review]

**Comment by breezykermo** [comment]
First minor thing that is my fault: could you rebase these changes on top of the latest release? There's a small fix to the build process that I think you found your way around there.

**Review by breezykermo — CHANGES_REQUESTED** [review]
This looks and works great, thanks! 

Just one change before we can merge this, as well as a couple of aesthetic suggestions. I've made two commits on a separate branch so that it's [easy to compare](https://github.com/Smoltbob/mtriage/compare/master...forensic-architecture:68e4a1f9a0d6).

The major change is to incorporate the downloading of weights into the main build process, which can be done by adding a line to partial.Dockerfile as shown. This will download the weights _at buildtime_, incorporating them as part of any mtriage image built with support for the `ProtestsPretrained` analyser. It loads the weights to the root folder, '/mtriage', so that those files baked into the Docker image aren't overwritten by the `--dev` flag or any of the other volumes that the mtriage outer layer (logic in [commands.py](commands.py)) mounts at runtime.

The other changes are just some aesthetic niceties around imports, mostly.

Thanks so much for your work on this! I'm excited to try it in practice during some of our workflows this week.

**Review by breezykermo — APPROVED** [review]

**Comment by breezykermo** [comment]
Just running a clean build and set of tests locally to confirm this all works as expected, then will merge if all goes well. Thanks again!

**Comment by breezykermo** [comment]
@Smoltbob how long would you expect inference to take on a CPU? I'm currently testing with ~10,000 frames on Intel I7 and nothing is produced yet after 20 minutes of running. 

Realising if inference for one frame takes closer to 0.5s than 0.1s this could be running for a while...

**Comment by Smoltbob** [comment]
Hi, so sorry for the delay! It takes 0.15sec on my i5-5200U, so that should be ok I think ? What I did notice when testing (including KerasPretrained) is that CPU in parallel doesn't seem to work. Could that be related ?

**Comment by breezykermo** [comment]
Ah I see. Yes I think you're right- I'll investigate the CPU in parallel problem.

**Comment by breezykermo** [comment]
Ah sorry- one other thing I just realised. Could you please PR to the 'release' branch? That way I can debug things before the changes flush through to master if something is up. Thanks so much, I'll merge as soon as the tests pass on the PR to release!

**Pull request closed without merging**
